include/lbann/layers/io/input/generic_input_layer.hpp

```diff
@@ -715,133 +726,84 @@ class generic_input_layer : public io_layer {
-    uint64_t validate_proc;
-    uint64_t validate_total;
-  };
-
   // reload state of IO from a checkpoint
   bool load_from_checkpoint_shared(persist& p) override {
     // save state of data readers from input layer
     data_reader_map_t::const_iterator it;
 
     it = this->m_data_readers.find(execution_mode::training);
     if ((it != this->m_data_readers.end()) && it->second) {
-      (it->second)->load_from_checkpoint_shared(p, "data_reader_training");
+      (it->second)->load_from_checkpoint_shared(p, execution_mode::training);
     }
     it = this->m_data_readers.find(execution_mode::testing);
     if ((it != this->m_data_readers.end()) && it->second) {
-      (it->second)->load_from_checkpoint_shared(p, "data_reader_testing");
-    }
-
-    // save our own state
-    // rank 0 reads the file
-    dataset_header header;
-    // Assume we are loading from a epoch end checkpoint
-    if (m_comm->am_trainer_master()) {
-      p.read_uint64(persist_type::train, "reader_train_processed",    &header.train_proc);
-      p.read_uint64(persist_type::train, "reader_train_total",        &header.train_total);
-      p.read_uint64(persist_type::train, "reader_test_processed",     &header.test_proc);
-      p.read_uint64(persist_type::train, "reader_test_total",         &header.test_total);
-      if(m_data_readers[execution_mode::validation] != nullptr){
-        p.read_uint64(persist_type::validate, "reader_validate_processed", &header.validate_proc);
-        p.read_uint64(persist_type::validate, "reader_validate_total",     &header.validate_total);
-      }
+      (it->second)->load_from_checkpoint_shared(p, execution_mode::testing);
     }
-
     it = this->m_data_readers.find(execution_mode::validation);
     if ((it != this->m_data_readers.end()) && it->second) {
-      (it->second)->load_from_checkpoint_shared(p, "data_reader_validation");
-    }
-    // TODO: assumes homogeneous hardware
-    // broadcast data from rank 0
-    MPI_Bcast(&header, sizeof(header), MPI_BYTE, 0, MPI_COMM_WORLD);
-    // set our fields
-    m_training_dataset.num_samples_processed()   = (long) header.train_proc;
-    m_training_dataset.total_samples()           = (long) header.train_total;
-    m_testing_dataset.num_samples_processed()    = (long) header.test_proc;
-    m_testing_dataset.total_samples()            = (long) header.test_total;
-    if(m_data_readers[execution_mode::validation] != nullptr){
-      m_validation_dataset.num_samples_processed() = (long) header.validate_proc;
-      m_validation_dataset.total_samples()         = (long) header.validate_total;
+      (it->second)->load_from_checkpoint_shared(p, execution_mode::validation);
     }
-    return true;
+
+    bool success = false;
+    std::string buf;
+    if (get_comm()->am_trainer_master()) {
+      success = read_cereal_archive<generic_input_layer>(*this, p, execution_mode::training, "_io.xml");
+      buf = create_cereal_archive_binary_string<generic_input_layer>(*this);
+   }
```

Comment:
This needs to do some sort of error handling. What if `am_trainer_master()` and `!success`? Then garbage is broadcast (presumably, since none of the state of `*this` will have been restored (again, presumably)).